File: IOI/2007/flood/solutions/robot.cpp
#include <stdio.h>
#include <string.h>

static char vis[2001][2001];

int sol = 0;

void dfs( int r, int c )
{
   static const char *command[] = { "up", "down", "left", "right" };
   static const int dr[] = { -1, 1, 0, 0 };
   static const int dc[] = { 0, 0, -1, 1 };

   ++sol;

   for ( int dir=0; dir<4; ++dir ) {
      int nr = r+dr[dir], nc = c+dc[dir];
      if ( vis[1000+nr][1000+nc] ) continue;
      
      vis[1000+nr][1000+nc] = 1;

      char reply[10];
      puts( command[dir] ); fflush( stdout );
      scanf( "%s", reply );
      if ( strcmp( reply, "ok" ) == 0 ) {
         dfs( nr, nc );
         puts( command[dir^1] ); fflush( stdout );
         scanf( "%s", reply );
      }
   }
}

int main()
{
   vis[1000+0][1000+0] = 1;
   dfs(0, 0);
   printf( "%d\n", sol );
   return 0;
}

File: IOI/2007/flood/solutions/robot.pas
program robot;

var
   vis : array[-1000..1000, -1000..1000] of boolean;
   sol : longint;

procedure dfs(r, c : longint);
const
   command : array[0..3] of string  = ('up', 'down', 'left', 'right');
   dr      : array[0..3] of longint = (-1, 1, 0, 0);
   dc      : array[0..3] of longint = (0, 0, -1, 1);
var
   dir    : longint;
   reply  : string;
   nr, nc : longint;
begin
   sol := sol+1;
   for dir:=0 to 3 do begin
      nr := r+dr[dir]; nc := c+dc[dir];
      if vis[nr, nc] then continue;

      vis[nr, nc] := true;
      writeln(command[dir]);
      readln(reply);
      if reply = 'ok' then begin
         dfs(nr, nc);
         writeln(command[dir xor 1]);
         readln(reply);
      end;
   end;
end;

begin
   vis[0, 0] := true;
   dfs(0, 0);
   writeln(sol);
end.
